bpo-29412: Fix indexError when parsing a header value ending unexpectedly

maxking opened the pull request:
When parsing a word, if the value is None, we should raise `HeaderParseError` instead of continuing silently. This helps fix the IndexError.

Part of the patch is from GH-6907, so thanks to @TyrannosourceExe.

<!-- issue-number: [bpo-29412](https://bugs.python.org/issue29412) -->
https://bugs.python.org/issue29412
<!-- /issue-number -->

TyrannosourceExe committed:
patched string index out of range error in get_word function of _header_value_parser.py and created tests in test__header_value_parser.py for CFWS.

maxking committed:
Raise HeaderParseError instead of continuing when parsing a word.

Also, modify the testcase.

maxking committed:
Undo code changed by mistake.

mangrisano reviewed (approved):
LGTM. Thank you for providing the test. It is always useful. :)

warsaw commented on Lib/email/_header_value_parser.py:
```diff
@@ -1360,6 +1360,9 @@ def get_word(value):
     else:
         leader = None
+    if not value:
+        raise errors.HeaderParseError("Expected 'atom' or 'quoted-string' "
```
I'd probably put a newline break after the open paren to see if you can avoid the implicit concatenation.

maxking replied:
Done.

warsaw commented on Lib/email/_header_value_parser.py:
```diff
@@ -1360,6 +1360,9 @@ def get_word(value):
         leader, value = get_cfws(value)
     else:
         leader = None
+    if not value:
```
Do you want to catch all false-y values or just `None`?

maxking replied:
I want to catch `None` and empty string `''` values, both of which could cause IndexError when running `value[0]`. It could also be caused by empty list, empty tuple etc etc, but I don't think `value` can be of those datatypes.

maxking committed:
Remove implicit concatenation

Remove implicit concatenation of strings and move the string
 to next line so that it can fit properly under 79 chars.
